auth/store.go

```diff
@@ -195,6 +200,76 @@ func (as *authStore) RoleAdd(r *pb.AuthRoleAddRequest) (*pb.AuthRoleAddResponse,
 	return &pb.AuthRoleAddResponse{}, nil
 }
 
+func (as *authStore) RoleGrant(r *pb.AuthRoleGrantRequest) (*pb.AuthRoleGrantResponse, error) {
+	tx := as.be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
+
+	_, vs := tx.UnsafeRange(authRolesBucketName, []byte(r.Name), nil, 0)
+	if len(vs) != 1 {
+		return nil, ErrRoleNotFound
+	}
+
+	role := &authpb.Role{}
+	err := role.Unmarshal(vs[0])
+	if err != nil {
+		plog.Errorf("failed to unmarshal a role %s: %s", r.Name, err)
+		return nil, err
+	}
+
+	var newPerm authpb.Permission_Type
+	if r.ReadPerm && r.WritePerm {
+		newPerm = authpb.READWRITE
+	} else if r.ReadPerm {
+		newPerm = authpb.READ
+	} else if r.WritePerm {
+		newPerm = authpb.WRITE
+	}
+
+	// TODO(mitake): clarify permission update rules
```

Comment:
At first I thought the rule would be complex but now I think it can be simple. The rules express what actions are allowed, not forbidden. So simple logical OR would be enough. e.g. if a prefix permission of read X and a key permission of Y are granted to a single role, simply granting both permission will be enough even if X contains Y. Just updating or appending would be enough.